velox/substrait/SubstraitToVeloxPlan.cpp

```diff
@@ -84,150 +84,226 @@ core::PlanNodePtr SubstraitVeloxPlanConverter::toVeloxPlan(
-      veloxGroupingExprs.emplace_back(fieldExpr);
-      outIdx += 1;
+      VELOX_CHECK(
+          groupingExpr.has_selection(), "Grouping is limited to be Field.");
+      constructExprs.emplace_back(exprConverter_->toVeloxExpr(
+          groupingExpr.selection(), constructInputType));
+      groupingOutIdx += 1;
     }
   }
 
-  // Parse measures to get Aggregation phase and expressions.
-  bool phaseInited = false;
-  core::AggregationNode::Step aggStep;
-  // Project expressions are used to conduct a pre-projection before
-  // Aggregation if needed.
-  std::vector<core::TypedExprPtr> projectExprs;
-  std::vector<std::string> projectOutNames;
-  std::vector<std::shared_ptr<const core::CallTypedExpr>> aggExprs;
-  aggExprs.reserve(aggRel.measures().size());
-
-  // Construct Velox Aggregate expressions.
-  for (const auto& sMea : aggRel.measures()) {
-    auto aggFunction = sMea.measure();
-    // Get the params of this Aggregate function.
-    std::vector<core::TypedExprPtr> aggParams;
-    auto args = aggFunction.args();
-    aggParams.reserve(args.size());
-    for (auto arg : args) {
-      auto typeCase = arg.rex_type_case();
-      switch (typeCase) {
-        case ::substrait::Expression::RexTypeCase::kSelection: {
-          aggParams.emplace_back(
-              exprConverter_->toVeloxExpr(arg.selection(), inputTypes));
-          break;
-        }
-        case ::substrait::Expression::RexTypeCase::kScalarFunction: {
-          // Pre-projection is needed before Aggregate.
-          // The input of Aggregatation will be the output of the
-          // pre-projection.
-          auto sFunc = arg.scalar_function();
-          projectExprs.emplace_back(
-              exprConverter_->toVeloxExpr(sFunc, inputTypes));
-          auto colOutName = substraitParser_->makeNodeName(planNodeId_, outIdx);
-          projectOutNames.emplace_back(colOutName);
-          auto outType = substraitParser_->parseType(sFunc.output_type());
-          auto aggInputParam =
-              std::make_shared<const core::FieldAccessTypedExpr>(
-                  toVeloxType(outType->type), colOutName);
-          aggParams.emplace_back(aggInputParam);
-          break;
-        }
-        default:
-          VELOX_NYI(
-              "Substrait conversion not supported for arg type '{}'", typeCase);
+  // Handle aggregations.
+  std::vector<std::string> aggFunctionNames;
+  aggFunctionNames.reserve(aggRel.measures().size());
+  std::vector<TypePtr> aggOutTypes;
+  aggOutTypes.reserve(aggRel.measures().size());
+
+  for (const auto& measure : aggRel.measures()) {
+    const auto& aggFunction = measure.measure();
+    std::string functionName = substraitParser_->findVeloxFunction(
+        functionMap_, aggFunction.function_reference());
+    aggFunctionNames.emplace_back(functionName);
+    aggOutTypes.emplace_back(toVeloxType(
+        substraitParser_->parseType(aggFunction.output_type())->type));
+    if (functionName == "avg") {
```

Comment:
Hardcoding logic for specific aggregate functions is hacky. Let's find a better way. Maybe lookup intermediate type in the registry, see if it is a struct (row) and if so look for the matching number of input columns and combine these into a struct.

Reply:
I fixed the comments of this PR at [#1851](https://github.com/facebookincubator/velox/pull/1851)